server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1315,15 +1315,185 @@ public void testTranslogWriter() throws IOException {
     }
 
+    public void testTranslogWriterFlushesOnWriteWhenBufferFull() throws IOException {
+        Path tempDir = createTempDir();
+        final TranslogConfig temp = getTranslogConfig(tempDir);
+        final TranslogConfig config = new TranslogConfig(temp.getShardId(), temp.getTranslogPath(), temp.getIndexSettings(),
+            temp.getBigArrays(), new ByteSizeValue(1, ByteSizeUnit.KB));
+
+        final Set<Long> persistedSeqNos = new HashSet<>();
+        final AtomicInteger writeCalls = new AtomicInteger();
+
+        final ChannelFactory channelFactory = (file, openOption) -> {
+            FileChannel delegate = FileChannel.open(file, openOption);
+            boolean success = false;
+            try {
+                // don't do partial writes for checkpoints we rely on the fact that the bytes are written as an atomic operation
+                final boolean isCkpFile = file.getFileName().toString().endsWith(".ckp");
+
+                final FileChannel channel;
+                if (isCkpFile) {
+                    channel = delegate;
+                } else {
+                    channel = new FilterFileChannel(delegate) {
+
+                        @Override
+                        public int write(ByteBuffer src) throws IOException {
+                            writeCalls.incrementAndGet();
+                            return super.write(src);
+                        }
+                    };
+                }
+                success = true;
+                return channel;
+            } finally {
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(delegate);
+                }
+            }
+        };
+
+        String translogUUID = Translog.createEmptyTranslog(
+            config.getTranslogPath(), SequenceNumbers.NO_OPS_PERFORMED, shardId, channelFactory, primaryTerm.get());
+
+        try (Translog translog = new Translog(config, translogUUID, new TranslogDeletionPolicy(),
+            () -> SequenceNumbers.NO_OPS_PERFORMED, primaryTerm::get, persistedSeqNos::add) {
+            @Override
+            ChannelFactory getChannelFactory() {
+                return channelFactory;
+            }
+        }) {
+            try (TranslogWriter writer = translog.createWriter(translog.currentFileGeneration() + 1)) {
+                int initialWriteCalls = writeCalls.get();
+                byte[] bytes = new byte[256];
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 1);
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 2);
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 3);
+                assertThat(persistedSeqNos, empty());
+                assertEquals(initialWriteCalls, writeCalls.get());
+
+                // This will fill the buffer and force a flush
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 4);
```

Comment:
Would be good to also include testing that the `readBytes` path works, for instance by randomly choosing to do readBytes here instead.